### ISSUE TITLE
Limiting default OpenMP threads 

### ISSUE THREAD

**Issue by mxz297** [dyninst/dyninst#693]
Currently SymtabAPI and ParseAPI will use whatever default number of OpenMP threads, which is likely to use all the available cores and hardware threads. This is not a desirable default behavior. 

I think that if the environment variable "OMP_NUM_THREADS" is set, we should respect that number and just do nothing. If "OMP_NUM_THREADS" is not set, how many threads should we use? My first impression is that we should use the small number between:

1. 32 threads: empirical evaluation shows that SymtabAPI and ParseAPI do not scale well beyond 32 threads.
2. half of the available hardware threads: we do not want to use all the available hardware resource by default. A question here is how do we get this number using c/c++ function call?

A separate question: should we provide programming interface to allow users to control the number of threads in case the user does not want to control the number of threads through "OMP_NUM_THREADS"?

@mwkrentel @hainest

**Comment by hainest** [comment]
My recommendation would be either

1. Through CMake using [ProcessorCount](https://cmake.org/cmake/help/latest/module/ProcessorCount.html).

2. Using [`std::thread::hardware_concurrency`](https://en.cppreference.com/w/cpp/thread/thread/hardware_concurrency)

(1) assumes that the build machine and the run machine have the same number of cores. (2) may or may not be implemented as completely as (1) for a given C++ standard library.

There still needs to be a heuristic involved since we can't just consume the entire machine, even if we map one thread per core. The linux kernel reports hyperthreads as unique cores (cf. `lscpu` or `/proc/cpuinfo`), so we need to avoid a 2x oversubscription on such platforms.

>should we provide programming interface to allow users to control the number of threads in case the user does not want to control the number of threads through "OMP_NUM_THREADS"?

I would hold off on this approach unless we are explicitly asked by a user to support it.

**Comment by mwkrentel** [comment]
It's a thorny issue, one without a clear, best answer.  Let's split
this into two questions and take them one at a time.

1. At a high level, what do you want openmp/symtab/dyninst client to
do?

2. What's the cleanest, most robust way to implement (1)?

I think (1) has a pretty clear answer: you want to allow the end user
to set the number of threads if s/he wants to, and if not, then
somebody somewhere should choose a sensible default.

And running 192 threads on po (power8, 48 cores, 4x HT) is not
sensible, at least not unless you explicitly ask for it.  So, I guess we
can blame openmp for their choice of a default.

----------

Now, let's look at solutions.  Let's identify 3 players: the openmp
runtime library, symtabAPI, and the dyninst client (top-level tool).

1. One solution is to do nothing, leave it up to the openmp layer.
There already is a way to set the number of threads:
omp_set_num_threads() or else OMP_NUM_THREADS.

The problem here is that someone who didn't know that symtab uses
threads may end up with an inappropriately high number of threads.
This violates the principle of "least shock and awe" to the naive
user.

(Another problem is that you may be running Dyninst as a tool inside
an openmp application.  It's awkward for OMP_NUM_THREADS to mean both
the app's threads and symtab's threads.)

2. Or, you could do nothing and leave it up to the dyninst client.
This is what hpcstruct does now.  We push this out to a command-line
option, we choose the default if the user doesn't, and we always
explicitly call omp_set_num_threads().

This is the same implementation as (1), except that here, if you get
the wrong number, then you blame the client instead of the end user.

It also requires every tool to know about this and expose the option
to the user.

----------

My takeaway from this is that Symtab should get involved.  Expose
something at the Symtab layer.  If the top-level client sets it, then
do that.  If not, then Symtab should choose a default.

Maybe an optional argument to openSymtab().  This would be cleaner if
C++ allowed named optional arguments.  Maybe another open function:
openSymtabWithThreads().

Whatever's convenient, but expose it at the Symtab level.
Same for ParseAPI.

Anyway, I'm a user and I'm asking.  I'm just not 100% sure of what the
best answer is.

----------

Btw, let me check my understanding.  The new elfutils 0.178 supports
parallel queries, but elfutils itself doesn't create any parallel
regions, is that right?

Anyone know if elfutils is planning on inserting openmp pragmas in
their code?

**Comment by mxz297** [comment]
> Btw, let me check my understanding. The new elfutils 0.178 supports
> parallel queries, but elfutils itself doesn't create any parallel
> regions, is that right?
> 
> Anyone know if elfutils is planning on inserting openmp pragmas in
> their code?

I don't think elfutils itself contains any openmp pragmas. We only make it thread-safe. 

@blue42u is this the case?

**Comment by blue42u** [comment]
Yes, Elfutils itself is only thread-safe. They use pthreads and atomics directly for synchronization.

**Comment by mwkrentel** [comment]
I looked up how the Intel Math Kernel Library (MKL) handles this.

https://software.intel.com/en-us/mkl-macos-developer-guide-techniques-to-set-the-number-of-threads

They have several knobs, but basically they added their own MKL layer
for controlling the number of threads.  In sort, the order of
precedence is:

 1. mkl_set_num_threads() function call.

 2. MKL_NUM_THREADS environ variable.

 3. Fall back to the OMP layer.

That would be a solution: add a method `symtab_set_num_threads()`.
For example:

```
symtab_set_num_threads(num);
Symtab::openFile(...);
```

Would that be too much violence to the API?

I'm still mulling over John's suggestion of "orphaned parallel
regions."

**Comment by mwkrentel** [comment]
Here's a sketch of a couple ways of handling this.

(1) The Intel MKL approach.

```
symtab_set_num_threads(num);
Symtab::openFile(...);
```

In this approach, the client explicitly tells Symtab how many threads
it wants.  If it doesn't, then Symtab should step in and enforce a
default value.

The advantage of an explicit call is that if it's not called, then
Symtab knows the client did not act.  No need to guess if
the client set `OMP_NUM_THREADS` or not.

----------

(2) John's "orphaned parallel regions" approach (if I have the correct
name).  The client runs:

```
omp_set_num_threads(num);

#pragma omp parallel
   {
      Symtab::openFile(...);
   }
```

In Symtab:

```
bool Symtab::openFile(...)
{
#pragma omp for
   for (...) {
      ...
   }
}
```

Again, the client sets the number of threads.  The main difference
is that if the client knows nothing about threads, then it defaults to
1 automatically.  But in (1), it's just as easy for Symtab to enforce
that itself.

----------

(3) Do nothing, what we have now.

Again, the client can set the number of threads by calling
`omp_set_num_threads()` itself (as hpcstruct does now).  But here, if
the client knows nothing about threads, then you're subject to the
"192 threads" shock and awe.

----------

I can accomodate any of the above, as we do now in (3).

But I suggest that Symtab (and ParseAPI) should get involved by some
means.

@jmellorcrummey  @n738cu

**Comment by jmellorcrummey** [comment]
If we use OpenMP tasking, then approach (2) is a bit different, which I think would be more comfortable:

My "orphaned parallel constructs" approach 



```
omp_set_num_threads(num);

#pragma omp parallel
   {
#pragma omp single
      Symtab::openFile(...);
   }
```

In SymtabAPI
```
bool Symtab::openFile(...)
{
...
   for (...) {
#pragma omp task
      ...
   }
}
```

The difference between this and what Mark wrote previously is that instead of having K parallel threads all call Symtab::openFile, which is awkward, only one thread calls SymtabAPI::openFile and the parallelism forks a number of tasks inside SymtabAPI that the worker threads in the parallel region all collaboratively execute.

As Mark said above when describing this orphaned parallel constructs model, all control of the parallelism is in the client.

The advantage of this tasking approach is that it makes it easier to allow tasks from multiple SymtabAPI invocations to run concurrently. For instance

(making this up)


```
omp_set_num_threads(num);

#pragma omp parallel
{
   #pragma omp single
   {
      Symtab::parseLineMaps(...);
      Symtab::dataflowAnalysis(...);
   }
}
```

Using the orphaned tasks construct strategy, tasks from parseLineMaps above could continue to execute beyond the end of the call to parseLineMaps and overlap with tasks from the call to dataflowAnalysis. 

You can't get that effect if all of the parallelism is from regions inside individual Dyninst APIs.

```
omp_set_num_threads(num);

#pragma omp parallel
{
   #pragma omp single
   {
   #pragma omp taskgroup
      {
         Symtab::parseLineMaps(...);
         Symtab::dataflowAnalysis(...);
      }
   }
}
```

The approach above says that tasks from parseLineMaps and dataflowAnalysis can execute in parallel. (They can execute beyond the end of their API call.)  At the end of the task group construct, we wait for them all to finish.
